certora/specs/TokensMorphoVaultV1Adapter.spec

```
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Copyright (c) 2025 Morpho Association
+
+using MorphoHarness as MorphoMarketV1;
+using MorphoVaultV1Adapter as MorphoVaultV1Adapter;
+using MetaMorpho as MetaMorpho;
+
+methods {
+    function _.extSloads(bytes32[]) external => NONDET DELETE;
+
+    function asset() external returns address envfree;
+    function lastUpdate() external returns uint64 envfree;
+    function liquidityData() external returns bytes envfree;
+
+    function MetaMorpho.asset() external returns address envfree;
+
+    function _.supply(MetaMorpho.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, bytes data) external with (env e)
+        => summarySupply(e, marketParams, assets, shares, onBehalf, data) expect (uint256, uint256) ALL;
+    function _.withdraw(MetaMorpho.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, address receiver) external with (env e) =>
+        summaryWithdraw(e, marketParams, assets, shares, onBehalf, receiver) expect (uint256, uint256) ALL;
+
+    function _.transfer(address, uint256) external => DISPATCHER(true);
+    function _.transferFrom(address, address, uint256) external => DISPATCHER(true);
+    function _.balanceOf(address) external => DISPATCHER(true);
+
+    // Required to avoid explicit linking for performance reasons.
+    function _.supplyShares(MorphoHarness.Id, address) external => DISPATCHER(true);
+    function _.deposit(uint256, address) external => DISPATCHER(true);
+    function _.withdraw(uint256, address, address) external => DISPATCHER(true);
+    function _.accrueInterest(MorphoHarness.MarketParams) external => DISPATCHER(true);
+
+    function _.borrowRate(MorphoHarness.MarketParams, MorphoHarness.Market) external => NONDET;
+    function _.canSendAssets(address) external => NONDET;
+    function _.canReceiveAssets(address) external => NONDET;
+}
+
+function summarySupply(env e, MetaMorpho.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, bytes data) returns (uint256, uint256)
+{
+    require (MetaMorpho.asset() == marketParams.loanToken, "safe require verified by Metamorpho's `MarketInteractions` and `ConsistentState` specifications");
+    MorphoMarketV1.supply(e, marketParams, assets, shares, onBehalf, data);
+    return (assets, shares);
```

```suggestion
    uint256 suppliedAssets;
    uint256 suppliedShares;
    (suppliedAssets, suppliedShares) = MorphoMarketV1.supply(e, marketParams, assets, shares, onBehalf, data);
    return (suppliedAssets, suppliedShares);
```
otherwise the return values are incorrect. This is especially problematic for `summaryWithdraw` because MM uses the return values of the `MORPHO.withdraw` in a non-trivial way